File: server_spring/src/main/java/de/werkelmann/rekote/server/CustomTrayIcon.java
package de.werkelmann.rekote.server;

import java.awt.AWTException;
import java.awt.Image;
import java.awt.MenuItem;
import java.awt.PopupMenu;
import java.awt.SystemTray;
import java.awt.TrayIcon;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.net.URL;

import javax.swing.ImageIcon;

public class CustomTrayIcon {

    public static void create() {
        if (!SystemTray.isSupported()) {
            return;
        }

        final PopupMenu popup = new PopupMenu();
        final TrayIcon trayIcon = new TrayIcon(getDefaultImage());
        final SystemTray tray = SystemTray.getSystemTray();

        MenuItem exitItem = new MenuItem("Exit");
        exitItem.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                System.exit(0);
            }
        });

        popup.add(exitItem);
        trayIcon.setPopupMenu(popup);

        try {
            tray.add(trayIcon);
        } catch (AWTException e) {
            System.out.println("Tray icon could not be added");
        }
    }

    static Image getDefaultImage() {
        return createImage("icon.gif", "tray icon");
    }

    private static Image createImage(String path, String description) {
        URL imageURL = CustomTrayIcon.class.getClassLoader().getResource(path);

        if (imageURL != null) {
            return (new ImageIcon(imageURL, description)).getImage();
        }
        throw new RuntimeException("Could not load tray icon");
    }
}

File: server_spring/src/main/kotlin/de/werkelmann/rekote/server/CustomTrayIcon.kt
package de.werkelmann.rekote.server

import de.werkelmann.rekote.server.dialog.PortDialog
import java.awt.*
import javax.swing.ImageIcon

object CustomTrayIcon {

    fun create() {
        if (!SystemTray.isSupported()) {
            return
        }

        val popup = PopupMenu()
        val trayIcon = TrayIcon(defaultImage)
        val tray = SystemTray.getSystemTray()

        val portDialogItem = MenuItem("Port")
        portDialogItem.addActionListener { PortDialog() }

        val exitItem = MenuItem("Exit")
        exitItem.addActionListener { System.exit(0) }

        popup.add(portDialogItem)
        popup.add(exitItem)
        trayIcon.popupMenu = popup

        try {
            tray.add(trayIcon)
        } catch (e: AWTException) {
            println("Tray icon could not be added")
        }

    }

    internal val defaultImage: Image
        get() = createImage("icon.gif", "tray icon")

    private fun createImage(path: String, description: String): Image {
        val imageURL = CustomTrayIcon::class.java.classLoader.getResource(path)

        if (imageURL != null) {
            return ImageIcon(imageURL, description).image
        }
        throw RuntimeException("Could not load tray icon")
    }
}
